### PR TITLE
RouterFS: implement RouterFileSystem

### LINKED ISSUE #3327: RouterFS: Support lakeFS via S3 gateway
As a first step, support lakeFS usage via the S3 gateway only. That means that the following configuration should work seamlessly:

```properties
fs.s3a.impl=RouterFileSystem
routerfs.mapping.s3a.1.replace=s3a://bucket/dir1/
routerfs.mapping.s3a.1.with=lakefs://repo/main/
fs.lakefs.impl=LakeFSFS
fs.s3a.access.key=<lakeFS access key>
fs.s3a.secret.key=<lakeFS access secret>
fs.s3a.endpoint=https://lakefs.example.com/routerfs
fs.s3a.path.style.access=true

routerfs.default.fs.s3a=S3AFileSystem # To handle other not lakeFS supported s3a URI paths
```
Parent issue: #2625

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
This PR includes a number of changes:

1. An implementation of the `RouterFileSystem` class' operations.
2. Upgrading our JUnit from version 4 to its latest stable 5 version, and also our Mockito version (if `mockito-inline` is necessary for some reason, please tell).
3. Adding Mockito support to JUnit using the Mockito extension for it.
4. Adding [project lombok](https://projectlombok.org/) to reduce the amount of code we write (if you don't like it for some reason please tell. This is optional).
5. `log4j.properties` file to configure our logger.
6. Some changes to the `PathMapper` class (including the renaming of `fromScheme` to `groupScheme`).

**How was this tested?**
Using the following Spark configurations and running a Spark application that uses them, RouterFS used lakeFS with (1) MinIO as it's backing object store and (2) S3 as it's backing store to successfully write the objects and their metadata:
 ```conf
fs.s3a.access.key=<lakeFS access key>
fs.s3a.secret.key=<lakeFS access secret>
fs.s3a.endpoint=https://lakefs.example.com
fs.s3a.path.style.access=true
fs.s3a.impl="io.lakefs.routerfs.RouterFileSystem"
fs.lakefs.impl="org.apache.hadoop.fs.s3a.S3AFileSystem"

routerfs.mapping.s3a.1.replace=s3a://bucket/dir1/
routerfs.mapping.s3a.1.with=lakefs://repo/main/
routerfs.default.fs.s3a=FilterFileSystem # To make sure that only paths of lakeFS will be handled by S3AFileSystem
```

Closes https://github.com/treeverse/lakeFS/issues/3327

### CODE REVIEW

File: src/main/resources/log4j.properties [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Set root logger level to INFO and its only appender to A1.
+log4j.rootLogger=INFO, A1
+log4j.appender.A1=org.apache.log4j.ConsoleAppender
+log4j.appender.A1.layout=org.apache.log4j.PatternLayout
+log4j.appender.A1.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
+
+log4j.logger.io..routerfs.*=INFO, A1
+
```
Comment (arielshaqed) [inline src/main/resources/log4j.properties]: This changes logging behaviour IIUC.  Isn't this something that the end-user should configure?

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 2]
```diff
@@ -199,14 +199,14 @@ public Path mapPath(Path origPath) {
...
-        LOG.trace("Converted % to % using path mapping %s", path, convertedPath,  pathMapping);
+        LOG.trace("Converted {} to {} using path mapping {}", path, convertedPath,  pathMapping);
         return new Path(convertedPath);
     }
 
     private PathMapping findAppropriatePathMapping(Path path) {
         Optional<PathMapping> appropriateMap = pathMappings.stream()
                 .filter(pm -> pm.isAppropriateMapping(path)).findFirst();
-        return appropriateMap.isPresent() ? appropriateMap.get() : null;
+        return appropriateMap.orElse(null);
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java:206]: This is a good change -- but it raises the question why not just change the return type to `Optional<PathMapping>` and handle presence that way down below when you use it?

(Not required, of course!)

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 3]
```diff
@@ -73,16 +74,8 @@ private void populatePathMappings(List<MappingConfig> mappingConfiguration) thro
...
-                priority2ConfigPair = new HashMap();
-                scheme2priority2ConfigPair.put(fromScheme, priority2ConfigPair);
-            }
-            List<MappingConfig> configPair = priority2ConfigPair.get(priority);
-            if (configPair == null) {
-                configPair = new ArrayList<>();
-                priority2ConfigPair.put(priority, configPair);
-            }
+            Map<Integer, List<MappingConfig>> priority2ConfigPair = scheme2priority2ConfigPair.computeIfAbsent(fromScheme, k -> new HashMap<>());
+            List<MappingConfig> configPair = priority2ConfigPair.computeIfAbsent(priority, k -> new ArrayList<>());
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java:79]: I don't understand.  `priority2ConfigPair` is a Map; how is this different from [`Map#computeIfAbsent`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-), introduced in Java **8**?

In general, this is a point _against_ using all-in-one packages such as Lombok: they tend to stick around long after the reasons for using them are gone.

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 4]
```diff
@@ -213,6 +206,7 @@ private PathMapping findAppropriatePathMapping(Path path) {
      * PathMapping represents a pair of mapping configurations, that define source and destination prefixes to
      * be replaced.
      */
+    @Getter
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java]: This annotation scares me.  IIUC, Lomboc will generate a getter method for _any_ private data member that I might add.  I.e. it changes the meaning of "private" for this class in a very strange way, and that will not be visible when adding new data members.

Auto-generated getters for single data members make sense, but I would rather have a library that does not let me do this whole-class annotation.

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 5]
```diff
@@ -233,7 +227,7 @@ public PathMapping(MappingConfig srcConfig, MappingConfig dstConfig, boolean def
             }
             this.fromScheme = srcConfig.getFromScheme();
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java]: `this.`?

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 6]
```diff
@@ -233,7 +227,7 @@ public PathMapping(MappingConfig srcConfig, MappingConfig dstConfig, boolean def
...
             if (!defaultMapping) {
-                if (srcConfig.getPriority() != dstConfig.getPriority()) {
+                if (srcConfig.getPriority().intValue() != dstConfig.getPriority().intValue()) {
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java]: The fact that we have this bug is a strong argument against using `Integer` to represent an optional integer in `MappingConfig`.  Also this is _also_ not correct, just in a different way: `intValue` will fail when `getPriority` is `null`.

So AFAICS we could _anyway_ keep priority as an `int` (maybe with a magic value -1 for "default").

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 7]
```diff
@@ -233,7 +227,7 @@ public PathMapping(MappingConfig srcConfig, MappingConfig dstConfig, boolean def
...
                     LOG.error("src and dst indices must match, cannot create PathMapping. src:"
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java]: Can we use a pattern (`{}`) here?

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 8]
```diff
@@ -279,11 +257,12 @@ public String toString() {
      * A class that represents a parsed Path mapping configurations supported by RouterFileSystem.
      * Mapping configuration form is: routerfs.mapping.${fromFsScheme}.${mappingIdx}.${replace/with}=prefix.
      */
+    @Getter
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/PathMapper.java:264]: A bit less scary here because it's a POJO, but still unpleasantly scary.

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 9]
```diff
@@ -48,8 +50,10 @@ public void initialize(URI name, Configuration conf) throws IOException {
             defaultToScheme = defaultFromScheme + DEFAULT_FS_SCHEME_SUFFIX;
         }
         conf.set("fs." + defaultToScheme + ".impl", defaultFsConf.getValue());
+        this.setConf(conf);
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: `this.`?

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 10]
```diff
@@ -120,7 +127,13 @@ public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
      */
     @Override
     public boolean rename(Path src, Path dst) throws IOException {
-        return false;
+        FileSystemPathTuple srcTuple = generateFSPathTuple(src);
+        FileSystemPathTuple dstTuple = generateFSPathTuple(dst);
+        if(!srcTuple.getFileSystem().equals(dstTuple.getFileSystem())) {
+            LOG.warn("Cannot rename paths from different FileSystems");
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]:
```suggestion
             LOG.warn("Cannot rename between different underlying FileSystems");
```
and ideally include the names of the filesystems.

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 11]
```diff
@@ -120,7 +127,13 @@ public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
      */
     @Override
     public boolean rename(Path src, Path dst) throws IOException {
-        return false;
+        FileSystemPathTuple srcTuple = generateFSPathTuple(src);
+        FileSystemPathTuple dstTuple = generateFSPathTuple(dst);
+        if(!srcTuple.getFileSystem().equals(dstTuple.getFileSystem())) {
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: I could not find a useful "equals" method for `FileSystem`s.  This may well cause issues!

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 12]
```diff
@@ -135,7 +148,8 @@ public boolean rename(Path src, Path dst) throws IOException {
      */
     @Override
     public boolean delete(Path f, boolean recursive) throws IOException {
-        return false;
+        FileSystemPathTuple tuple = generateFSPathTuple(f);
+        return tuple.getFileSystem().delete(tuple.getPath(), recursive);
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java:168]: Let's mark with a bug that this fails when recursively deleting across a filesystem.  It probably does not matter, but let's document it.

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 13]
```diff
@@ -160,7 +175,13 @@ public FileStatus[] listStatus(Path f) throws FileNotFoundException, IOException
      */
     @Override
     public void setWorkingDirectory(Path new_dir) {
-
+        try {
+            FileSystemPathTuple tuple = generateFSPathTuple(new_dir);
+            this.workingDirectory = tuple.getPath();
+            tuple.getFileSystem().setWorkingDirectory(tuple.getPath());
+        } catch (IOException e) {
+            LOG.error("Failed setting a working directory with exception:\n{}", e.getLocalizedMessage());
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java:199]: I don't understand: why do we want to swallow an exception here?

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 14]
```diff
@@ -196,10 +218,13 @@ public boolean mkdirs(Path f, FsPermission permission) throws IOException {
...
-//        return fs.getFileStatus(mappedPath);
-        return null;
+        FileSystemPathTuple tuple = generateFSPathTuple(f);
+        return tuple.getFileSystem().getFileStatus(f);
+    }
+
+    private FileSystemPathTuple generateFSPathTuple(Path p) throws IOException {
+        Path mappedPath = this.pathMapper.mapPath(p);
+        FileSystem fs = mappedPath.getFileSystem(getConf());
+        return new FileSystemPathTuple(fs, mappedPath);
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: Thinking about it: I don't understand how this works after the current working directory is changed.  If I look at `"./path/to/obj"`, we will not know **which** underlying filesystem to open, so it doesn't matter that I changed directories in that FS.

File: src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
+package io.lakefs.routerfs.dto;
+
+import lombok.Getter;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+@Getter
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java]: [@Value](https://projectlombok.org/features/Value)?  (You could also get value types from other projects if you don't like this specific Value type...).

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Please note that I commented on each commit separately; it might be easier to read the review in the same way.

Big request is to handle working directories, which I am not sure will work with the current implementation.

File: src/main/resources/log4j.properties (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply src/main/resources/log4j.properties]: You're right! It was supposed to be under the tests dir.
thanks

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 5 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java]: can you elaborate?

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 7 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java]: Done

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 3 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java:79]: I don't really understand... 
This **is** [Map#computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-).
I used the functional style that Java 8 offers instead of the "if" statements from before.
 It has nothing to do with Lombok.

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 4 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java]: I see what you mean.
As I saw it, this class would probably not contain any private fields that we wouldn't want to expose, but I can use the annotation on each field separately.

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 6 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java]: I'll do a mix of the two: the constructor will still receive an `Integer` as an argument so we can check for `null` values instead of the client classes passing the default number. But the class's field type will be `int` and it will be initialized by default to -1 and will get changed if the passed arg isn't null

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 9 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: oops

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 13 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java:199]: The overridden method doesn't throw exceptions

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 11 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: You're right.
Will compare the FSs' URIs

File: src/main/java/io/lakefs/routerfs/PathMapper.java [hunk 16]
```diff
@@ -68,21 +69,14 @@ private void loadMappingConfig(Configuration conf) throws InvalidPropertiesForma
      *
      * @param mappingConfiguration the configurations to create path mapping from
      */
-    private void populatePathMappings(List<MappingConfig> mappingConfiguration) throws InvalidPropertiesFormatException {
+    private List<PathMapping> populatePathMappings(List<MappingConfig> mappingConfiguration) throws InvalidPropertiesFormatException {
+        List<PathMapping> pathMappings = new ArrayList<>();
         Map<String, Map<Integer, List<MappingConfig>>> scheme2priority2ConfigPair = new HashMap<>();
         for (MappingConfig conf : mappingConfiguration) {
-            String fromScheme = conf.getFromScheme();
+            String fromScheme = conf.getGroupScheme();
```
Comment (talSofer) [inline src/main/java/io/lakefs/routerfs/PathMapper.java]:
```suggestion
            String groupScheme = conf.getGroupScheme();
```

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 8 above)
Comment (talSofer) [reply src/main/java/io/lakefs/routerfs/PathMapper.java:264]: Why are we using the @Getter annotation in different ways? that is, in Mapping config the annotation is at the class level as opposed to PathMapping where we use it at the data member level?

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 17]
```diff
@@ -48,8 +56,13 @@ public void initialize(URI name, Configuration conf) throws IOException {
             defaultToScheme = defaultFromScheme + DEFAULT_FS_SCHEME_SUFFIX;
         }
         conf.set("fs." + defaultToScheme + ".impl", defaultFsConf.getValue());
+        setConf(conf);
         super.initialize(name, conf);
-        this.pathMapper = new PathMapper(conf, defaultFromScheme, defaultToScheme);
+        if(this.pathMapper == null) {
```
Comment (talSofer) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java:61]:
```suggestion
        if (this.pathMapper == null) {
```

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 18]
```diff
@@ -14,23 +16,29 @@
...
     private static final String DEFAULT_FS_CONF_PATTEREN = "^routerfs\\.default\\.fs\\.(?<fromScheme>[-a-z0-9_]*)";
     private static final String DEFAULT_FS_SCHEME_REGEX_GROUP_NAME = "fromScheme";
     private static final String DEFAULT_FS_SCHEME_SUFFIX = "-default";
     private static final String DEFAULT_FS_CONF_PREFIX = "routerfs.default.fs";
 
     private PathMapper pathMapper;
+    private Path workingDirectory;
+    private URI uri;
+
+    public RouterFileSystem(PathMapper pathMapper) {
```
Comment (talSofer) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java:32]: Do we need this constructor for testing purposes?

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 19]
```diff
@@ -196,10 +234,23 @@ public boolean mkdirs(Path f, FsPermission permission) throws IOException {
...
+
+    private FileSystemPathTuple generateFSPathTuple(Path p) throws IOException {
+        p = createAbsolutePath(p);
+        Path mappedPath = this.pathMapper.mapPath(p);
+        FileSystem fs = mappedPath.getFileSystem(getConf());
+        return new FileSystemPathTuple(fs, mappedPath);
+    }
+
+    private Path createAbsolutePath(Path p) {
+        if(p.isAbsolute()) {
```
Comment (talSofer) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]:
```suggestion
        if (p.isAbsolute()) {
```

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 20]
```diff
@@ -135,7 +158,14 @@ public boolean rename(Path src, Path dst) throws IOException {
      */
     @Override
     public boolean delete(Path f, boolean recursive) throws IOException {
-        return false;
+        FileSystemPathTuple tuple = generateFSPathTuple(f);
+        /*
+        Potential Bug:
+            If `recursive == true`, the recursive process may cause the underlying filesystems to change.
+            For example, if "s3a://bucket/dir1/" is mapped to "fs1://bla/", and "s3a://bucket/dir1/dir2/" is mapped to
+            "fs2://blah/" then the recursive process will fail if the path points to "s3a://bucket/" or "s3a://bucket/dir1/"
```
Comment (talSofer) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java:166]: Thanks for documenting this. That was one of my concerns. did you test it with our Sonatype Spark app that (I think) is doing some deletes during the write process?

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+        this.routerFileSystemUnderTest = new RouterFileSystem(this.mockPathMapper);
+        Configuration configuration = generateConfiguration();
+        lenient().when(this.mockPath.getFileSystem(configuration)).thenReturn(this.mockFileSystem);
+        this.routerFileSystemUnderTest.initialize(uri, configuration);
+    }
+
+    @Test
+    public void testGetUriSuccessful() {
+        URI uri = routerFileSystemUnderTest.getUri();
+        assertEquals(uri.getPath(), uri.getPath());
```
Comment (talSofer) [inline src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java]: Is this test a WIP? we are comparing an object to iteself

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    @Test
+    public void testOpen(@Mock FSDataInputStream mockFSDataInputStream) throws IOException {
+        int bufferSize = 3;
```
Comment (talSofer) [inline src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java]: you can make this a constant since your are using it in multiple tests

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+        FSDataInputStream result = this.routerFileSystemUnderTest.open(PATH, bufferSize);
+        assertEquals(mockFSDataInputStream, result);
+        verify(this.mockFileSystem, times(1)).open(this.mockPath, bufferSize);
+    }
+
+    @Test
+    public void testCreate(@Mock FsPermission mockPermission, @Mock Progressable mockProgressable, @Mock FSDataOutputStream mockFSDataOutputStream) throws IOException {
+        int bufferSize = 3;
+        short sh = 1000;
+        long lng = 20000;
```
Comment (talSofer) [inline src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java]:
```suggestion
        long lng = 20000L;
```

File: src/test/resources/log4j.properties [hunk 24]
```diff
@@ -0,0 +1,8 @@
+# Set root logger level to INFO and its only appender to A1.
+log4j.rootLogger=INFO, A1
+log4j.appender.A1=org.apache.log4j.ConsoleAppender
+log4j.appender.A1.layout=org.apache.log4j.PatternLayout
+log4j.appender.A1.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
+
+log4j.logger.io..routerfs.*=INFO, A1
```
Comment (talSofer) [inline src/test/resources/log4j.properties]: The two dots after io are intentional here? if so, what does this mean? :)

**Review by talSofer — COMMENTED** [review]
Neat! looks good. 
One important question - how was this tested? 
Also thanks for improving the style!

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 18 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java:32]: Unfortunately... If only we could implement dependency injection according to our rules and not Hadoop's...

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java (hunk 21 above)
Comment (Jonathan-Rosenberg) [reply src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java]: 😅​ it was but then just forgot to delete it...
Thanks

File: src/test/resources/log4j.properties (hunk 24 above)
Comment (Jonathan-Rosenberg) [reply src/test/resources/log4j.properties]: 100% typo
thanks!

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 8 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/PathMapper.java:264]: You can see the discussion about this [here](https://github.com/treeverse/hadoop-router-fs/pull/7/commits/b6147be0ff46efaf1b764749e4852c72bfb47e79)...
Basically, @arielshaqed raised a possible leak of information from this class, so I changed to more specific getter annotations, while `MappingConfig` is a POJO thus less of a risk...

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 20 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java:166]: Nope. Not aware of this spark app... What is it?

**Comment by Jonathan-Rosenberg** [comment]
@talSofer I'm in the middle of the process of testing it. Basically, I'm configuring Router FS as specified [here](https://github.com/treeverse/lakeFS/issues/3327) with two types of backing stores (lakeFS will be backed by MinIo and s3a will be backed by S3), so two different writes and reads should succeed on two different locations.

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java (hunk 20 above)
Comment (talSofer) [reply src/main/java/io/lakefs/routerfs/RouterFileSystem.java:166]: https://github.com/treeverse/lakeFS/tree/master/test/spark/app

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 3 above)
Comment (arielshaqed) [reply src/main/java/io/lakefs/routerfs/PathMapper.java:79]: Sorry, thought I saw Lombok's `computeIfAbsent`.

File: src/main/java/io/lakefs/routerfs/RouterFileSystem.java [hunk 25]
```diff
@@ -14,23 +16,29 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+@NoArgsConstructor
 public class RouterFileSystem extends FileSystem {
 
-    public static final Logger LOG = LoggerFactory.getLogger(RouterFileSystem.class);
+    private static final Logger LOG = LoggerFactory.getLogger(RouterFileSystem.class);
     private static final String DEFAULT_FS_CONF_PATTEREN = "^routerfs\\.default\\.fs\\.(?<fromScheme>[-a-z0-9_]*)";
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/RouterFileSystem.java]: Not part of this PR, but
```suggestion
    private static final String DEFAULT_FS_CONF_PATTERN = "^routerfs\\.default\\.fs\\.(?<fromScheme>[-a-z0-9_]*)";
```
❓

File: src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java [hunk 26]
```diff
@@ -0,0 +1,13 @@
+package io.lakefs.routerfs.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+@Value
+@AllArgsConstructor
+public class FileSystemPathTuple {
```
Comment (arielshaqed) [inline src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java]: Not a fan of this name.  Specifically "tuple" is typically the generic term.  This one _is_ a `FileSystemAndPath` (and also a pair).  The word "tuple" adds mystery and I had to look up this definition to understand what it is.

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+        this.routerFileSystemUnderTest.initialize(uri, configuration);
+    }
+
+    @Test
+    public void testOpen(@Mock FSDataInputStream mockFSDataInputStream) throws IOException {
+        when(this.mockFileSystem.open(this.mockPath, BUFFER_SIZE)).thenReturn(mockFSDataInputStream);
+        FSDataInputStream result = this.routerFileSystemUnderTest.open(PATH, BUFFER_SIZE);
+        assertEquals(mockFSDataInputStream, result);
+        verify(this.mockFileSystem, times(1)).open(this.mockPath, BUFFER_SIZE);
+    }
```
Comment (arielshaqed) [inline src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java:49]: I don't understand this test: it verifies that a single `open` method calls the default `open` method.  We're going to test the routing logic later, right?

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Test
+    public void testCreate(@Mock FsPermission mockPermission, @Mock Progressable mockProgressable, @Mock FSDataOutputStream mockFSDataOutputStream) throws IOException {
+        short sh = 1000;
+        long lng = 20000L;
```
Comment (arielshaqed) [inline src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java]: I don't understand these variable names.  Can we name them for what they are, rather than for their types?

File: src/main/java/io/lakefs/routerfs/PathMapper.java (hunk 6 above)
Comment (arielshaqed) [reply src/main/java/io/lakefs/routerfs/PathMapper.java]: Maybe just 2 constructors, then?

**Review by arielshaqed — APPROVED** [review]
Thanks!

I am a bit unsure about whether or not we want to have `this.` prefixes before every instance variable.  I don't want them, but what I want least of all is inconsistency (e.g. on line 45) in using them.

(Personally I believe this prefix is **bad** practice, as it is not enforced by the language; indeed, Java specifically has scoping rules that allow it.  It is pretty much guaranteed to be inconsistent, e.g. l. 45.  So it puts needless burden onto the human programmers: I have to make sure to write `this.` each time, but I can never rely on it and have to check each lookup to determine whether or not it is a member access.)

We can decide on a team style (or decide on some external style guide), but then we should abide by it.

File: src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java (hunk 26 above)
Comment (Jonathan-Rosenberg) [reply src/main/java/io/lakefs/routerfs/dto/FileSystemPathTuple.java]: not a fan of using `And` in class names, but we can bridge the gap with `FileSystemPathPair` (pair is a 2-tuple 😉​)

File: src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java (hunk 27 above)
Comment (Jonathan-Rosenberg) [reply src/test/java/io/lakefs/routerfs/RouterFileSystemTest.java:49]: This test makes sure that when `open` gets called on `RouterFileSystem`, the underlying FileSystem's (whichever it is) `open` method gets called, and that the expected `FSDataInputStream` is returned.
The `PathMapper` is where the magic happens- this is where the mapping occurs, then we just call (the already routed path's) `mappedPath.getFileSystem(getConf())` to retrieve the correct FileSystem.

To your question, the routing should be tested in the `PathMapper` tests, and integration tests (testing the classes' workflow with mocks/real filesystem implementations) should also be created.

**Review by talSofer — APPROVED** [review]
Cool! 
Can you please add the description of the testing you did to the pr description?
